### ISSUE TITLE
Intermittent 70 second gaps in metrics

### ISSUE THREAD

**Issue by leejones** [frodenas/stackdriver_exporter#51]
Hi! First, thank you for creating/sharing this tool! It's been very helpful to us on our journey with Prometheus and Google Cloud.

I'm hoping someone can help with an issue we've been running into. I noticed intermittent gaps in our Prometheus metrics that are pulled from the stackdriver-exporter. This occurs about 10 times each day and the gaps are consistently 70 seconds long.

The gaps correlate to Prometheus scrape timeouts of the exporter as well as log messages from the exporter that indicate timeouts during requests to the Stackdriver API, for example:

```
time="2019-04-24T20:09:52Z" level=error msg="Error retrieving Time Series metrics for descriptor `pubsub.googleapis.com/subscription/num_retained_acked_messages_by_region`: Get https://monitoring.googleapis.com/v3/projects/redacted/timeSeries?alt=json&filter=metric.type%3D%22pubsub.googleapis.com%2Fsubscription%2Fnum_retained_acked_messages_by_region%22&interval.endTime=2019-04-24T20%3A09%3A42.119800239Z&interval.startTime=2019-04-24T20%3A04%3A42.119800239Z: net/http: request canceled (Client.Timeout exceeded while awaiting headers)" source="monitoring_collector.go:182"
time="2019-04-24T20:09:52Z" level=error msg="Error retrieving Time Series metrics for descriptor `pubsub.googleapis.com/subscription/num_retained_acked_messages`: Get https://monitoring.googleapis.com/v3/projects/redacted/timeSeries?alt=json&filter=metric.type%3D%22pubsub.googleapis.com%2Fsubscription%2Fnum_retained_acked_messages%22&interval.endTime=2019-04-24T20%3A09%3A42.119800239Z&interval.startTime=2019-04-24T20%3A04%3A42.119800239Z: net/http: request canceled (Client.Timeout exceeded while awaiting headers)" source="monitoring_collector.go:182"
time="2019-04-24T20:09:52Z" level=error msg="Error while getting Google Stackdriver Monitoring metrics: Get https://monitoring.googleapis.com/v3/projects/redacted/timeSeries?alt=json&filter=metric.type%3D%22pubsub.googleapis.com%2Ftopic%2Funacked_bytes_by_region%22&interval.endTime=2019-04-24T20%3A09%3A42.087937084Z&interval.startTime=2019-04-24T20%3A04%3A42.087937084Z: net/http: request canceled (Client.Timeout exceeded while awaiting headers)" source="monitoring_collector.go:138"
```

 Here's some details from our environment:

* Prometheus
  * version: 2.7.1
  * scrape interval: ~10 seconds~ 60 seconds (for the stackdriver-exporter target)
  * scrape timeout: 10 seconds
* stackdriver-exporter
  * version: 0.6.0
  * timeout: 10 seconds

Are these timeouts and/or metric gaps something that other people have seen? Any other ideas on how to troubleshoot this?

**Comment by leejones** [comment]
I recently realized that our scrape interval for the stackdriver-exporter is set to 60s (instead of our 10s default for everything else). That helps me understand why there's a 60-70 second gap instead of a 10-20 second one.

I'm still confused about the Stackdriver API timeouts themselves. Is anyone else seeing those when using the stackdriver-exporter? I'm running it from a container with GKE.

**Comment by leejones** [comment]
I discovered that the Google Cloud Console has metrics for each Google API. I looked at the metrics for the Stackdriver Monitoring API (used by the exporter) and found that these gaps corresponded to API degradation on Google's side (e.g elevated latency and error rates). I asked Google Support about this and didn't get an answer about why this happens but they did give some general guidance for the client side:

>  try breaking up each request into a set of smaller requests by setting label filters. Otherwise, they can retry timeouts with exponential backoff.

I guess there's something that could be done on the client side to improve this a little bit, but I don't know that it's worth it. The API seems to meeting its [99.95% SLO](https://cloud.google.com/stackdriver/sla) so I don't know if it's worth trying to optimize beyond that on the client side. I'm ok with a few intermittent gaps in my current use case.
